Full width dropdowns for Vanilla "meganav" navigation

bartaz opened the pull request:
## Done

This PR adds a new style of top navigation dropdowns that are full-width containers.
Mobile version still uses the same "sliding" navigation lists of items, but desktop dropdowns will contain a separate content, with full support for grid and Vanilla components or any custom HTML.

NOTE: this feature is part of bigger work in progress for upstreaming meganav styling and is only meant to be used in the context of meganav on canonical.com or other websites.



Fixes https://warthogs.atlassian.net/browse/WD-12784

## QA

**NOTE**: the functionality is only about dropdown container, the **content of the dropdown is not part of this feature**. For demo we have some test content that uses grid.

- Open [demo](https://vanilla-framework-5213.demos.haus/docs/examples/patterns/navigation/dropdown-full-width?theme=light)
- Full width dropdowns on "home page" top navigation
https://vanilla-framework-5213.demos.haus/docs/examples/patterns/navigation/dropdown-full-width?theme=light
- Full width dropdowns on reduced navigation:
https://vanilla-framework-5213.demos.haus/docs/examples/patterns/navigation/reduced?theme=light
- Test both examples above 
  - make sure they open and close as expected
  - make sure it interacts well with search
  - make sure switching between dropdowns works
  - make sure on small screens sliding nav works as before (having it's own separate content)
- Review updated documentation:
  - https://vanilla-framework-5213.demos.haus/docs/patterns/navigation#full-width-dropdowns
  - (make sure you view on screen large enough to have embedded example show the dropdowns)

### Check if PR is ready for release

If this PR contains Vanilla SCSS code changes, it should contain the following changes to make sure it's ready for the release:

- [x] PR should have one of the following labels to automatically categorise it in release notes:
  - `Feature 🎁`, `Breaking Change 💣`, `Bug 🐛`, `Documentation 📝`, `Maintenance 🔨`.
- [x] Vanilla version in `package.json` should be updated relative to the [most recent release](https://github.com/canonical/vanilla-framework/releases/latest), following semver convention:
  - if CSS class names are not changed it can be bugfix relesase (x.x.**X**)
  - if CSS class names are changed/added/removed it should be minor version (x.**X**.0)
  - see the [wiki for more details](https://github.com/canonical/vanilla-framework/wiki/Release-process#pre-release-tasks)
- [x] Any changes to component class names (new patterns, variants, removed or added features) should be listed on the [what's new page](https://github.com/canonical/vanilla-framework/blob/main/releases.yml).


## Screenshots

<img width="1771" alt="image" src="https://github.com/user-attachments/assets/3c973e6d-ed90-4990-991e-84e40d6c93c7">

webteam-app commented:
### [<img src='https://assets.ubuntu.com/v1/6baef514-ubuntu-circle-of-friends-large.svg' height=32 width=32> Demo</img>](https://vanilla-framework-5213.demos.haus)
### [<img src='https://assets.ubuntu.com/v1/e512b0e2-jenkins.svg' height=32 width=32> Jenkins </img>](https://jenkins.canonical.com/webteam/job/start-demo/4450)
### [<img src='https://assets.ubuntu.com/v1/7144ec6d-logo-jaas-icon.svg' height=32 width=32> demos.haus </img>](https://demos.haus)

jmuzina commented on scss/_patterns_navigation.scss:
```diff
@@ -763,6 +767,56 @@ $spv-navigation-logo-bottom-position: 0.125rem; // 2px when 1rem is 16px
+    }
+
+    &.is-collapsed .p-navigation__dropdown-content {
+      transform: translateY(-101%);
```
Just checking, is this `tranform: translateY(-101%)` here for animation purposes or something else? It's not immediately clear from the code so it may be worth adding a comment

bartaz replied:
Yes, it's for animation. I'll add something.

jmuzina commented on templates/docs/examples/patterns/navigation/dropdown-full-width.html:
This file needs to be added to `navigation/combined.html`.

bartaz commented:
> One other comment I'll keep separate from my review, since it seems possibly out of scope - I notice there are some minor wrapping issues at specific screen sizes, particularly while desktop nav is displayed:

Thanks @pastelcyborg. Yes, the wrapping is kind of known issue that would not happen in real use case. We have a `$breakpoint-navigation-threshold` variable that you adjust based on amount of content in the nav in your project.

This example we have here has a lot more than usual navigations and I don't want to adjust the Vanilla default to accomodate that. The problem is, we can't easly have this setting adjusted per example, because it's part of compile step, so we would need a separate Vanilla build for this specific example.

bartaz commented:
> Since the component requires JS anyway, it might be nice to move to a model where we calculate the metrics of the nav discreetly and apply `$breakpoint-navigation-threshold` automatically, so the user doesn't need to do anything on their end. Definitely out of scope for this PR, though.

Responsive breakpoints are built on compile-time using the SCSS variables. We can't change that from JS. To move responsiveness to JS we would have to rework how mobile/desktop styles are applied, as it would have to happen not via media queries, but via JS (class name change?). Definitely out of scope .

petesfrench commented on templates/docs/examples/patterns/navigation/_script-sliding.js:
```diff
@@ -106,30 +106,46 @@ const initNavigationSliding = () => {
   toggles.forEach(function (toggle) {
```
Some logic responsible for closing the navigation items (ex. removing and adding the 'is-collapsed' class) shouldn't be included in the event listen for the buttons themselves as this isn't the sole way to close the dropdowns

bartaz replied:
Good point!

bartaz commented:
I just realised I messed up something in JS and sliding navigation only works for first level of nested menu, but not for second one. Need to dig out of it.

bartaz commented on templates/docs/examples/patterns/navigation/_full-width-dropdown.jinja:
```diff
@@ -0,0 +1,87 @@
+<div class="p-navigation__dropdown is-full-width is-collapsed"  data-level="1" id="{{ id }}" aria-hidden="true">
+  <div class="p-navigation__dropdown-content-mobile">
+    <ul class="p-list u-no-margin">{# TODO: use p-list temporarly, we will later have more specific styling class name for it #}
+      <li class="p-navigation__item--dropdown-close" id="{{ id }}-back">
```
I get it, but these are what they are now, and changing them at this point may result in unexpected breaking changes for anyone using this.

The structure of the navigation (especially combined with the meganav dropdowns) is so deep now, that BEM convention barely holds up (although it's probably a sign that this should be split into smaller components).

I get your point, but changing those now would be a breaking change.

bartaz commented on templates/docs/examples/patterns/navigation/_full-width-dropdown.jinja:
```diff
@@ -0,0 +1,87 @@
+      </li>
+    </ul>
+  </div>
+  <div class="p-navigation__dropdown-content">
```
Fair point. I assumed it would be a "default" (desktop content), and mobile as an alternative.

Also "small" "large" could be confusing here, as navigation does not use our standard breakpoints, but has mobile/desktop view based on amount of elements in the nav, so is switching between 2 views at a breakpoint that is configurable, and may not be aligned with rest of Vanilla.

bartaz replied:
Renamed to `p-navigation__dropdown-content--sliding` and `p-navigation__dropdown-content--full-width` for consistency and clarity.

bartaz commented on templates/docs/examples/patterns/navigation/_script-sliding.js:
```diff
@@ -106,30 +106,46 @@ const initNavigationSliding = () => {
           target.setAttribute('aria-hidden', 'false');
+
+          // only animate the dropdown if menu is not open, otherwise just switch the visible one
+          if (!navigation.classList.contains('has-menu-open')) {
```
It's based on how browser handles rendeing (and how the CSS transitions are triggered).

The `is-collapsed` class triggers a CSS transition. But, it if gets applied together with the rest of the state change that makes the element appear (display none is removed), the transition would not trigger (browser just paints it all together).

To trigger the CSS transition `is-collapsed` class name needs to be applied separately, in the next paint event. So that first, the elements get's rendered in "collapsed" state, and (a fraction of a milisecond later) the `is-collapsed` class will be applied and trigger CSS transition.

Don't know how to get it clearer from the code without an extensive comment.

bartaz commented on templates/docs/examples/patterns/navigation/_script-sliding.js:
```diff
@@ -106,30 +106,46 @@ const initNavigationSliding = () => {
+            toggle.parentNode.classList.remove('is-active');
+            toggle.parentNode.parentNode.classList.remove('is-active');
+            navigation.classList.remove('has-menu-open');
+          }, 100);
```
Yes, this actually should align with the transition time from CSS. But because we don't have a "JavaScript library" exported with Vanilla it's not standarised in JS in any way. But I can make it a const for the purposes of this script for clarity.

bartaz commented on templates/docs/examples/patterns/navigation/_script-sliding.js:
```diff
@@ -190,6 +206,22 @@ const initNavigationSliding = () => {
   }
+
+  // throttle util (for window resize event)
+  var throttle = function (fn, delay) {
```
We could, but Vanilla not being a JS library, we don't really have a JS dependency, or a shared JS files. We also have to make sure that all that is needed for certain example gets included (rendered) in the example HTML.

So far, we've been coping these, but I agree that we are getting to a point where it would be worth having a look at organising the Vanilla JS a bit better.

bartaz replied:
Why isn't it part of browser native JS api yet? 🤷‍♂️

bartaz commented on templates/docs/examples/patterns/navigation/_script-sliding.js:
```diff
@@ -106,30 +106,46 @@ const initNavigationSliding = () => {
       if (target) {
-        const isNested = target.parentNode.parentNode.classList.contains('p-navigation__dropdown');
-
+        const isNested = !target.closest('.p-navigation__dropdown');
```
In this particular case, it's not really a JS hook, we are checking if specific dropdown is on top level or not, so it kind of relies on checking parent CSS class name.

bartaz replied:
I'd want to avoid refactoring the whole script (that was not written for this feature) as part of this PR.

Because it's not about just this place here. The whole script would have to be significantly rewritten to remove the references of CSS class names or specific parent nodes.

bartaz commented on templates/docs/examples/patterns/navigation/_script-sliding.js:
```diff
@@ -82,7 +82,7 @@ const initNavigationSliding = () => {
     if (target.closest) {
       if (!target.closest('.p-navigation--sliding, .p-navigation--reduced')) {
-        resetToggles();
+        closeAll();
```
Good call.

bartaz commented on scss/_patterns_navigation.scss:
```diff
@@ -798,6 +852,7 @@ $spv-navigation-logo-bottom-position: 0.125rem; // 2px when 1rem is 16px
 
   .p-navigation--sliding.has-menu-open,
   .p-navigation--reduced.has-menu-open {
+    box-shadow: rgba(0, 0, 0, 0.5) 0px 0px 0px 100vh;
```
Default Vanilla overlay seems too dark, but will confirm.

bartaz replied:
Updated

petesfrench commented on templates/docs/examples/patterns/navigation/_full-width-dropdown.jinja:
```diff
@@ -0,0 +1,87 @@
+  <div class="p-navigation__dropdown-content-mobile">
+    <ul class="p-list u-no-margin">{# TODO: use p-list temporarly, we will later have more specific styling class name for it #}
+      <li class="p-navigation__item--dropdown-close" id="{{ id }}-back">
+        <button aria-controls="{{ id }}" class="p-navigation__link js-back">
```
```suggestion
        <button aria-controls="{{ id }}" class="p-navigation__link js-back-button">
```
Could we make this 'js-back-button' to align with 'js-menu-button' and 'js-search-button'?

bartaz replied:
Sure, I haven't been touching much of the sliding nav things, but it doesn't hurt to do small fixes like this right away.

bartaz replied:
Done.

jmuzina commented:
I'm having trouble opening the reduced dropdown on mobile size, is this expected?

[Screencast from 07-23-2024 09:36:37 AM.webm](https://github.com/user-attachments/assets/f8a4c85a-64b2-4bf6-a8d8-c7c842ab01f4)

lyubomir-popov commented:
the only thing is the top one should be muted, so darker text:
![image](https://github.com/user-attachments/assets/3496a322-89c1-4846-89d2-fe2597c6f478)

bartaz commented:
> I'm having trouble opening the secondary dropdown on mobile size, is this expected?

Good catch @jmuzina. Fixed!

jmuzina commented on templates/docs/patterns/navigation/index.md:
```diff
@@ -86,6 +86,21 @@ When used on websites with more complex information archutecture there may be a
+
+To use them add `.is-full-width` class name to `.p-navigation__dropdown` element. Inside of it you need two containers for full-width desktop dropdown content `.p-navigation__dropdown-content--full-width` and mobile sliding navigation `.p-navigation__dropdown-content--sliding` content. The mobile content will be displayed on small screens only and desktop content on larger screens (based on the v ...
+
+<div class="embedded-example"><a href="/docs/examples/patterns/navigation/dropdown-full-width" class="js-example">View example of the full-width dropdown pattern</a></div>
```
There seems to be some whitespace control issues in the rendered markup here:

![image](https://github.com/user-attachments/assets/adb05cd0-6e72-42e7-bbaf-e1c01b22780e)

Though I think this is OK considering this is marked as work-in-progress and there is other work dependent on this; I'd rather not hold the PR up over this

bartaz replied:
Parts of this are generated from macro, so I guess macro defines the indentation, not the parent HTML.

I suppose the idea to have external formatter would be good in long term, not to rely on how HTML source is structured.

jmuzina replied:
cc @pastelcyborg related to our previous conversation about external formatter

petesfrench commented:
It looks like the 'setFocus' function isn't working and thus keyboard navigation is broken. I addressed this in my PR, [here](https://github.com/canonical/canonical.com/pull/1310/files#diff-2deebc3e9a704771e51d3329f9080b6768ea5d16b5c7185270f65acf189c730e)

bartaz commented:
> It looks like the 'setFocus' function isn't working and thus keyboard navigation is broken. I addressed this in my PR, [here](https://github.com/canonical/canonical.com/pull/1310/files#diff-2deebc3e9a704771e51d3329f9080b6768ea5d16b5c7185270f65acf189c730e). Basically it doesn't pick up the `ul` as it is nested

Good catch @petesfrench. I think I managed to address it (it turned out to be a bit more complicated than your fix, but it was a good direction). I also introduced the `js-dropdown-nav-list` class name that needs to be added to every list that needs to be made focusable/unfocusable.

petesfrench commented:
@bartaz Interesting, I have done the same thing. It might be worth doing a review of the code I write and the code in Vanilla to see if we can align them. So you everything can work together out the box as it does on c.com.

bartaz commented:
Thanks everyone for reviews and tips!
